getting-started/src/main/kotlin/com/google/cloud/kotlin/emojify/EmojifyController.kt

```diff
@@ -0,0 +1,149 @@
+/*
+ * Copyright 2018 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.kotlin.emojify
+
+import com.google.cloud.storage.Bucket
+import com.google.cloud.storage.Storage
+import com.google.cloud.vision.v1.Likelihood
+import com.google.cloud.vision.v1.FaceAnnotation
+import com.google.cloud.vision.v1.AnnotateImageRequest
+import com.google.cloud.vision.v1.ImageAnnotatorClient
+import com.google.cloud.vision.v1.ImageSource
+import com.google.cloud.vision.v1.Image
+import com.google.cloud.vision.v1.Feature
+import com.google.cloud.vision.v1.Feature.Type
+import org.springframework.beans.factory.annotation.Value
+import org.springframework.web.bind.annotation.GetMapping
+import org.springframework.web.bind.annotation.RequestParam
+import org.springframework.web.bind.annotation.RestController
+import java.awt.Polygon
+import java.awt.image.BufferedImage
+import java.io.ByteArrayOutputStream
+import java.io.InputStream
+import java.nio.channels.Channels
+import javax.imageio.ImageIO
+import java.util.logging.Logger
+
+enum class Emoji {
+    JOY, ANGER, SURPRISE, SORROW, NONE
+}
+
+// Returns best emoji based on detected emotions likelihoods
+fun bestEmoji(annotation: FaceAnnotation): Emoji {
+    val emotionsLikelihood = listOf(Likelihood.VERY_LIKELY, Likelihood.LIKELY, Likelihood.POSSIBLE)
+    val emotions = mapOf(
+        Emoji.JOY to annotation.joyLikelihood,
+        Emoji.ANGER to annotation.angerLikelihood,
+        Emoji.SURPRISE to annotation.surpriseLikelihood,
+        Emoji.SORROW to annotation.sorrowLikelihood
+    )
+    for (likelihood in emotionsLikelihood) { // In this order: VERY_LIKELY, LIKELY, POSSIBLE
+        for (emotion in emotions) { // In this order: JOY, ANGER, SURPRISE, SORROW (https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/map-of.html)
+            if (emotion.value == likelihood) return emotion.key // Returns emotion corresponding to likelihood
+        }
+    }
+    return Emoji.NONE
+}
+
+class EmojifyResponse(
+    val objectPath: String? = null,
+    val emojifiedUrl: String? = null,
+    val status: Int = 200,
+    val errorMessage: String? = null
+)
+
+@RestController
+class EmojifyController(@Value("\${storage.bucket.name}") val bucketName: String, val storage: Storage, val vision: ImageAnnotatorClient) {
+
+    val log = Logger.getLogger(EmojifyController::class.java.name)
```

Comment:
You may want to look at putting some of your static variables in a companion object so they are reused (this may not apply for Spring, but worth checking out): https://stackoverflow.com/questions/34416869/idiomatic-way-of-logging-in-kotlin

Reply:
Sure!